src/objects/BatchedMesh.ts

```diff
@@ -0,0 +1,425 @@
+
+    let size = Math.sqrt(this._maxGeometryCount * 4) // 4 pixels needed for 1 matrix
+    size = MathUtils.ceilPowerOfTwo(size)
+    size = Math.max(size, 4)
+
+    const matricesArray = new Float32Array(size * size * 4) // 4 floats per RGBA pixel
+    const matricesTexture = new DataTexture(matricesArray, size, size, RGBAFormat, FloatType)
+
+    this._matricesArray = matricesArray
+    this._matricesTexture = matricesTexture
+    this._matricesTextureSize = size
+
+    this._customUniforms.batchingTexture.value = this._matricesTexture
+    this._customUniforms.batchingTextureSize.value = this._matricesTextureSize
+  }
+
+  _initShader() {
+    const material = this.material as Material
+    const currentOnBeforeCompile = material.onBeforeCompile
+    const customUniforms = this._customUniforms
+
+    material.onBeforeCompile = function onBeforeCompile(parameters, renderer) {
+      // Is this replacement stable across any materials?
+      parameters.vertexShader = parameters.vertexShader
+        .replace('#include <morphcolor_vertex>', '#include <morphcolor_vertex>\n' + batchingbaseVertex)
+        .replace('#include <skinning_pars_vertex>', '#include <skinning_pars_vertex>\n' + batchingParsVertex)
+        .replace('#include <skinnormal_vertex>', '#include <skinnormal_vertex>\n' + batchingnormalVertex)
+        .replace('#include <skinning_vertex>', '#include <skinning_vertex>\n' + batchingVertex)
+
+      for (const uniformName in customUniforms) {
+        parameters.uniforms[uniformName] = customUniforms[uniformName]
+      }
+
+      //currentOnBeforeCompile.call(this, parameters, renderer)
+    }
+    material.customProgramCacheKey = () =>
+      Object.entries(customUniforms)
+        .map(([k, v]) => (k + v.value) as string)
+        .reduce((x, y) => x + y)
+    material.defines = material.defines || {}
+    material.defines.BATCHING = false
+    material.needsUpdate = true
+  }
+
+  getGeometryCount() {
+    return this._geometryCount
+  }
+
+  getVertexCount() {
+    return this._vertexCount
+  }
+
+  getIndexCount() {
+    return this._indexCount
+  }
+
+  applyGeometry(geometry: BufferGeometry) {
+    // @TODO: geometry.groups support?
+    // @TODO: geometry.drawRange support?
+    // @TODO: geometry.mortphAttributes support?
+
+    if (this._geometryCount >= this._maxGeometryCount) {
+      // @TODO: Error handling
+    }
+
+    if (this._geometryInitialized === false) {
+      for (const attributeName in geometry.attributes) {
+        const srcAttribute = geometry.getAttribute(attributeName)
+        const { array, itemSize, normalized } = srcAttribute
+
+        const dstArray = new (array.constructor as Float32ArrayConstructor)(this._maxVertexCount * itemSize)
+        const dstAttribute = new (srcAttribute.constructor as any)(dstArray, itemSize, normalized) as BufferAttribute
+
+        // TODO: add usage in @types/three
+        // @ts-ignore
+        dstAttribute.setUsage(srcAttribute.usage)
```

Comment:
This looks to not be typed https://threejs.org/docs/#api/en/core/BufferAttribute.usage.